File: leetcode/dp/tree/0337_house_robber_iii.cpp
// LC 337. House Robber III | Medium (DP on Tree)
#include <algorithm>
using namespace std;
struct TreeNode { int val; TreeNode *left, *right; TreeNode(int x) : val(x), left(nullptr), right(nullptr) {} };

class Solution {
    pair<int,int> dfs(TreeNode* n) {
        if (!n) return {0, 0};
        auto [lr, ls] = dfs(n->left);
        auto [rr, rs] = dfs(n->right);
        return {n->val + ls + rs, max(lr, ls) + max(rr, rs)};
    }
public:
    int rob(TreeNode* root) { auto [r, s] = dfs(root); return max(r, s); }
};

File: leetcode/dp/tree/0337_house_robber_iii.py
# LC 337. House Robber III | Medium (DP on Tree)
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val; self.left = left; self.right = right

class Solution:
    def rob(self, root):
        def dfs(node):
            if not node: return (0, 0)
            l = dfs(node.left); r = dfs(node.right)
            rob_it = node.val + l[1] + r[1]
            skip_it = max(l) + max(r)
            return (rob_it, skip_it)
        return max(dfs(root))
